### ISSUE TITLE
Too many permissions needed to create a VM

### ISSUE THREAD

**Issue by Avengerist** [Telmate/terraform-provider-proxmox#784]
I found that in version 2.9.14 should be much more permissions to create a VM compared to version 2.9.11 (not tested 2.9.13)
Also, the error doesn't say about which permissions you need.

Error in version 2.9.14 if you are using **Token**:
```
╷
│ Error: user username@pve has valid credentials but cannot retrieve user list, check privilege separation of api token
│
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on main.tf line 25, in provider "proxmox":
│   25: provider "proxmox" {
│
```
Error in version 2.9.14 if you are using **Password:**
```
╷
│ Error: Request cancelled
│
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on main.tf line 25, in provider "proxmox":
│   25: provider "proxmox" {
│
│ The plugin.(*GRPCProvider).ConfigureProvider request was cancelled.
╵
```

Compared to version 2.9.11 It needs additional permissions:
- **Sys.Audit**
- **Sys.Modify**
- **Sys.Console**

**Comment by Budlee** [comment]
Added those permissions when using an API token and am still seeing the same error

**Comment by hestiahacker** [comment]
I can confirm that this issue was introduced in 2.9.13 (and also affect 2.9.14).

**Comment by hestiahacker** [comment]
After a LOT of digging, I believe I've tracked down why **Sys.Audit** is necessary as of 2.9.13.

- provider.go [calls](https://github.com/Telmate/terraform-provider-proxmox/blob/master/proxmox/provider.go#L225) client.GetUserPermissions (introduced in [refactoring of proxmox-api-go](https://github.com/Telmate/terraform-provider-proxmox/pull/654)))
  - client.go [accesses the permissions of a specific userID](https://github.com/Telmate/proxmox-api-go/blob/03f4e428f6c62f9496d388d1735777b6c832cba2/proxmox/client.go#L1632), which [requires Sys.Audit](https://ge.hax0rbana.org/api2/json/access/users?full=1) permission on /access.

This appears to have been added by @mleone87 in commit [9b75cbb](https://github.com/Telmate/proxmox-api-go/commit/9b75cbb854cf0abc44477082bad67f27a7d09105#diff-0a45dc0a0e482bac94ff577e2b0e0562efb7d438c489eadff7138771ed2b2c0fR1600).  All of this, ironically, is to enable the error messages to tell the user what permissions are needed, and to attempt to not start making changes if they are going to fail later.

The Sys.Audit permissions is only required to get the permission of other users.  To get the permissions of yourself, the `userId=` parameter can be omitted and the API will return the list of your permissions. In this case, we do want to get the permissions of ourselves, however client.go::GetUserPermissions() is written to get the permissions of a generic userId with no concept of getting the permissions of the logged in user.

I propose the following to remove the need for Sys.Audit:
1. add GetMyPermissions() to client.go
2. call this new function in proxmox/provider.go instead of calling GetUserPermissions

Obviously testing will be required to ensure there are not any other API calls that require Sys.Audit.

If anyone is willing to make this change, I can repeat this deep dive into the code to try to track down why Sys.Modify and Sys.Console are now required. I just want to make sure that spending more hours tracing through code, diffing commits, checking API docs, and the like is actually going to result in improvements at some point.

**Comment by TuncTaylan** [comment]
Even with Sys.Audit, Sys.Modify and Sys.Console I have the same issue with the versions 2.9.13 and 2.9.14. It works with 2.9.11 though.

**Comment by TuncTaylan** [comment]
Any update?

**Comment by hestiahacker** [comment]
> Any update?

Unfortunately not from me. Life has interrupted my plan to set up a test server so I can try to reproduce the issue of Sys.Audit, Sys.Modify and Sys.Console permissions being given but still running into this error.

It is still something I'm trying to do (because I need a proper test server anyway), and I'm less busy this month than last month, so at least I'm heading in the right direction there. Can't say I have an ETA for you though. :confused:

**Comment by SahorSs** [comment]
Something has changed?

**Comment by hestiahacker** [comment]
This may be related to #687, which is also on my list of things to investigate and hopefully fix. Between these two tickets, I should be able to extensively document what permissions are required for password authentication and what permissions are needed when using a token.

I only have one Proxmox server for testing and I'm not too keen on changing the version of that for this test, but I'll make sure I include what Proxmox version I'm testing with in each testing session.

**Comment by mohsen-saeidi** [comment]
Unfortunately, our cloud provider is refusing to give us permission for Sys.Modify and Sys.Console. It was working with version 2.9.10 and now we have to update the provider version. Can you somehow make the permissions configurable?

**Comment by Avengerist** [comment]
Did you find a solution to fix this?

**Comment by hestiahacker** [comment]
OK, I have more information on precisely what is required, but I don't have any workarounds.

Before I get into the details, I have a few suggested paths forward:

- **Short term**: add an option which will skip the permission checks and just try to do everything as if it has permission in proxmox. Something like a boolean called performPermissionChecks with a default value of true.
- **Long term**: Try to find a way to reduce the number of permissions required. This may mean determining exactly what permissions a specific terraform file is going to require (e.g. if it's not using CloudInit, don't check/error out if this permission is missing)
  - As I [mentioned before](https://github.com/Telmate/terraform-provider-proxmox/issues/784#issuecomment-1646517496), I believe the requirement for Sys.Audit can be avoided; I haven't looked into the rest

If we can get some agreement on these two things, we can open a ticket to add the feature to not perform permission checks, and at least one ticket in the [proxmox-api-go](https://github.com/Telmate/proxmox-api-go) repo, as I believe most of the code that will need changed is going to be over there.

With the first ticket, it's possible that we'll find out that skipping the checks is not sufficient, and we'll need to skip certain steps that we do now. If that turns out to be the case, then we'll have to dig into the long term solution to get this resolved.

**At this point, I'd like to try to get a consensus on whether these two paths forward seem like they're an approach the people following this ticket want. Because the first step at resolving this issue is agreeing on the requirements and next steps.**

# 2.9.11
I've done extensive testing and found that Sys.Audit, Sys.Console, and Sys.Modify are not actually required at all in v2.9.11. Here are the minimum required permissions as of version 2.9.11:

- Proxmox custom role (path: /vms)
  - VM.Allocate (required)
  - VM.Audit (required)
  - VM.Clone (required)
  - VM.Config.CPU (required)
  - VM.Config.Disk (required)
  - VM.Config.HWType (required)
  - VM.Config.Memory (required)
  - VM.Config.Network (required)
  - VM.Config.Options (required)
- Proxmox custom role (paths: /storage/local-zfs)
  - Datastore.AllocateSpace (required)

It's possible there may be some terraform files that can be deployed with fewer permissions, but for the typical use case of cloning a base VM and configuring it, these are the minimum.

Here's the terraform definitions for this test:
```
terraform {
  required_providers {
    proxmox = {
      #source = "registry.example.com/telmate/proxmox"
      #version = ">=1.0.0"
      source = "Telmate/proxmox"
      version = "=2.9.11"
      #version = "=3.0.1-rc1"
    }
  }
  required_version = ">= 0.14"
}
resource "proxmox_vm_qemu" "server" {
  name              = "test.example.com"
  target_node       = "ra"
  clone             = "debian-12"
  full_clone        = true
  os_type           = "cloud-init"
  cores             = 1
  sockets           = "1"
  cpu               = "host"
  memory            = 512
  scsihw            = "virtio-scsi-pci"
  bootdisk          = "virtio0"
  disk {
    size            = "20G"
    type            = "virtio"
    cache           = "writeback"
    storage         = "local-zfs"
  }
#  disks {
#    virtio {
#      virtio0 {
#        disk {
#          size            = 20
#          cache           = "writeback"
#          storage         = "local-zfs"
#        }
#      }
#    }
#  }
  network {
    model           = "virtio"
    bridge          = "vmbr0"
  }

  # Cloud Init Settings
  # Reference: https://pve.proxmox.com/wiki/Cloud-Init_Support
  ipconfig0 = "ip=192.168.22.222/22,gw=192.168.22.1"
  nameserver = "192.168.22.100 192.168.23.100"
  sshkeys = file("${path.root}/test.pub")
```

# 3.0.1-rc1
Then I changed the terraform file to use the 3.0.1-rc1 version of the provider, commented out the `disk` section and uncommented the `disks` block.

## Round 1
Attempting to deploy resulted in the following errors:

```
 Error: Invalid provider configuration
│ 
│ Provider "registry.terraform.io/telmate/proxmox" requires explicit configuration. Add a provider block to the root module and configure the provider's required
│ arguments as described in the provider documentation.
│ 
╵
╷
│ Error: Plugin did not respond
│ 
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on <empty> line 0:
│   (source code not available)
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ConfigureProvider call. The plugin logs may contain more details.
```

If I add the permission for Sys.Audit, the error message becomes much more helpful.
```
│ Error: Invalid provider configuration
│ 
│ Provider "registry.terraform.io/telmate/proxmox" requires explicit configuration. Add a provider block to the root module and configure the provider's required
│ arguments as described in the provider documentation.
│ 
╵
╷
│ Error: permissions for user/token issue687@pam are not sufficient, please provide also the following permissions that are missing: [Datastore.AllocateSpace Datastore.Audit Pool.Allocate VM.Allocate VM.Audit VM.Clone VM.Config.CDROM VM.Config.CPU VM.Config.Cloudinit VM.Config.Disk VM.Config.HWType VM.Config.Memory VM.Config.Network VM.Config.Options VM.Migrate VM.Monitor VM.PowerMgmt]
│ 
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on <empty> line 0:
│   (source code not available)
```

Of these, the following are new and were not previously required:

- Datastore.Audit
- Pool.Allocate
- VM.Config.CDROM
- VM.Config.Cloudinit
- VM.Migrate
- VM.Monitor
- VM.PowerMgmt

To understand what these are, we can go to the [Proxmox admin guide](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#pveum_permission_management), which offers the explanations I've pasted below.

- Datastore.Audit: view/browse a datastore
- Pool.Allocate: create/modify/remove a pool
- VM.Config.CDROM: eject/change CD-ROM
- VM.Config.Cloudinit: modify Cloud-init parameters
- VM.Migrate: migrate VM to alternate server on cluster
- VM.Monitor: access to VM monitor (kvm)
- VM.PowerMgmt: power management (start, stop, reset, shutdown, …)

I'm not sure why VM.Monitor and Pool.Allocate are required, but the rest of these seem pretty reasonable. I could understand making VM.Config.CDROM and VM.Config.Cloudinit optional, as not everyone is going to want to use CloudInit. I did verify that the list of permissions in the error message did not change if I remove the cloudinit section of my terraform definition.

This left me more wondering how on Earth it was setting up Cloudinit (which uses an emulated CDROM) without those permissions. I can confirm that it **was** doing it though!

## Round 2
After adding the permissions above, I then got the same error message, but with different permissions that were missing. The additional ones this time (along with what Proxmox's documentation says they do) were:

- Sys.Console: console access to node
- Sys.Modify: create/modify/remove node network parameters

However there were additional requirements that were harder to notice. The error also kept listing all the VM.* permissions unless I changed the scope to be "/" instead of "/vms". The same was required for Datastore.AllocateSpace. Even using "/storage" was not sufficient, it had to be "/".

## Round 3
After adding all of the above permissions, the deployment started, but then failed with a 403 error. So after all those up-front checks, there were still some things that were needed that were not in that list! Here's the error message:

```
│ Error: error updating VM: 403 Permission check failed (/vms/101, VM.Config.Cloudinit), error status: {"data":null}
```

I'm not sure what the problem was there, but after removing the permissions to specific paths, and keeping only the ones that covered "/", it deployed without any issue. My best guess is that it has something to do with "propagation" which is whether the user gets permissions on entities that don't exist yet (like /vms/101) just because they have the necessary permission on (/vms or /).

**Comment by JamborJan** [comment]
Thanks a lot for your hard work, @hestiahacker !

I assume that most folks have found a workaround. At least I have (using a root-like user, no API key). So, I would opt for the direct path to the proper solution. I guess nobody has too much free time for a temporary solution, plus I think it might end up in users thinking the main issue is solved and feeling too safe. Especially disabling the checks and ending up in unknown states where part of the infrastructure is created and part not is quite frustrating.

**Comment by hestiahacker** [comment]
It looks like there is at least [one person](https://github.com/Telmate/terraform-provider-proxmox/issues/784#issuecomment-1900661914) who is unable to update to the latest version.

I fear you may be right that these permissions are likely now required and not doing the check is not going to resolve the issue, but there is a chance. I expect changing the code back to where it doesn't require the additional permissions is going to take quite a long time and I don't want to leave people hanging if there's a way we can get them the same functionality that old versions had (which did not do proactive permission checks).

**Comment by mohsen-saeidi** [comment]
Thanks for your hard work @hestiahacker, fortunately after some discussions with our cloud provider, they gave me a specific API token with the required permissions, so I can use the provider again :)

**Comment by JamborJan** [comment]
Are you using a provider managed Proxmox service @mohsen-saeidi ? If yes, which one are you using? I was not thinking about such an option. In my little thinking Proxmox was all about selfhoating today :)

**Comment by mohsen-saeidi** [comment]
@JamborJan Unfortunately I am not sure if I can mention their name because of privacy and data protection :) but yes we are a private company and our cloud provider is hosting all of our infrastructure via Proxmox.

**Comment by Tinyblargon** [comment]
@hestiahacker Checking permissions dynamically was merged upstream https://github.com/Telmate/proxmox-api-go/issues/308

But after attempting to use this in multiple places, I've realized some shortfalls of it.

**Comment by mrkhachaturov** [comment]
# 1. Create Role

```
pveum role add TerraformProv -privs "Datastore.AllocateSpace Datastore.Audit Pool.Allocate Sys.Audit Sys.Console Sys.Modify VM.Allocate VM.Audit VM.Clone VM.Config.CDROM VM.Config.Cloudinit VM.Config.CPU VM.Config.Disk VM.Config.HWType VM.Config.Memory VM.Config.Network VM.Config.Options VM.Migrate VM.Monitor VM.PowerMgmt"
```

# 2. Create User
# - note the command balks at special characters in the terminal
# - ended up resetting to strong password via GUI
```
pveum user add terraform-prov@pve --password <password>
```

# 3. Associate Role & User
```
pveum aclmod / -user terraform-prov@pve -role TerraformProv
```

API Token
From the proxmox GUI, create an API Token for the terraform-prov user.

ensure "privilege separation" for your API Token is disabled or terraform will error out later on:
![image](https://github.com/Telmate/terraform-provider-proxmox/assets/105451445/72e133cb-752d-4779-acb1-76ced7da42d2)


[Source](https://ronamosa.io/docs/engineer/LAB/proxmox-terraform)

**Comment by maksimsamt** [comment]
> @hestiahacker Checking permissions dynamically was merged upstream [Telmate/proxmox-api-go#308](https://github.com/Telmate/proxmox-api-go/issues/308)
> 
> But after attempting to use this in multiple places, I've realized some shortfalls of it.

@Tinyblargon,
Is it possible to make checking permissions functionality ( [Telmate/proxmox-api-go#308](https://github.com/Telmate/proxmox-api-go/issues/308)) optional? True/false or no permissions check for some user, as for example now is for root@pam.
In my case, because of this functionality it is not possible to use user token with privilege separation and with permissions for the specific pool and with built-in roles. Unfortunately, it is not good practice to set acl permissions for / (root pool).

**Comment by maksimsamt** [comment]
My workaround to get working using user's token with privilege separation and pool's fine-grained permissions.

System details:
Proxmox VE 8.2.2
Terraform v1.8.4
terraform-provider-proxmox v3.0.1-rc2

- On PVE side (execution in pve terminal):

  ```bash
  # create custome role
  pveum role add TerraformProv -privs "\
  VM.Allocate \
  VM.Audit \
  VM.Clone \
  VM.Config.CDROM \
  VM.Config.CPU \
  VM.Config.Cloudinit \
  VM.Config.Disk \
  VM.Config.HWType \
  VM.Config.Memory \
  VM.Config.Network \
  VM.Config.Options \
  VM.Migrate \
  VM.Monitor \
  VM.PowerMgmt"

  # create user
  pveum user add terraform-prov@pve --password <changeme>

  # create token with privilege separation and no expire
  pveum user token add terraform-prov@pve terraform-prov-token -privsep 1 --expire 0

  # create a new resource pool
  pveum pool add terraform-prov-pool -comment "Hashicorp Terraform Provisioning Test Pool (user terraform-prov@pve)"

  # add the desired datastores (e.g. local-lvm) to "terraform-prov-pool" resource pool:
  pvesh set /pools/terraform-prov-pool --storage local-lvm

  # Note, for user if need you can grant full Administrator or PVEVMAdmin a.o. role for /
  # add specific roles for terraform-prov-pool (user: terraform-prov@pve):
  pveum acl modify \
  /pool/terraform-prov-pool/ \
  -user 'terraform-prov@pve' \
  -roles "\
  TerraformProv \
  PVEPoolUser \
  PVEDatastoreUser"

  # Note, Separated privileges: The token needs to be given explicit access with ACLs. Its effective permissions are calculated by intersecting user and token permissions.
  # add specific roles for terraform-prov-pool (token: terraform-prov@pve!terraform-prov-token):
  pveum acl modify \
  /pool/terraform-prov-pool/ \
  -token 'terraform-prov@pve!terraform-prov-token' \
  -roles "\
  TerraformProv \
  PVEPoolUser \
  PVEDatastoreUser"

  # add PVEDatastoreAdmin role for iso storage (e.g. /storage/lvm, user: terraform-prov@pve):
  pveum acl modify \
  /storage/lvm \
  -user 'terraform-prov@pve' \
  -role PVEDatastoreAdmin

  # add PVEDatastoreAdmin role for iso storage (e.g. /storage/lvm, token: terraform-prov@pve!terraform-prov-token):
  pveum acl modify \
  /storage/lvm \
  -token 'terraform-prov@pve!terraform-prov-token' \
  -role PVEDatastoreAdmin

  # add PVESDNUser role for localnetwork (user: terraform-prov@pve):
  pveum acl modify \
  /sdn/zones/localnetwork \
  -user 'terraform-prov@pve' \
  -role PVESDNUser

  # add PVESDNUser role for localnetwork (token: terraform-prov@pve!terraform-prov-token):
  pveum acl modify \
  /sdn/zones/localnetwork \
  -token 'terraform-prov@pve!terraform-prov-token' \
  -role PVESDNUser

  # add PVESysAdmin for / (user: terraform-prov@pve):
  pveum acl modify / -user 'terraform-prov@pve' -role PVESysAdmin

  # add PVESysAdmin for / (token: terraform-prov@pve!terraform-prov-token):
  pveum acl modify / -token 'terraform-prov@pve!terraform-prov-token' -role PVESysAdmin

  # optional (templates-pool should exist): add view and clone templates permission (user: terraform-prov@pve):
  pveum acl modify /pool/templates-pool/ \
  -user 'terraform-prov@pve' \
  -role PVETemplateUser

  # optional (templates-pool should exist): add view and clone templates permission (token: terraform-prov@pve!terraform-prov-token):
  pveum acl modify /pool/templates-pool/ \
  -token 'terraform-prov@pve!terraform-prov-token' \
  -role PVETemplateUser
  ```

- `terraform-provider-proxmox` modification:

  <https://github.com/Telmate/terraform-provider-proxmox/blob/master/proxmox/provider.go#L196>

  Modify `minimumPermissions` as per below example (keep only `Sys.Audit` permission):

  ![image](https://github.com/Telmate/terraform-provider-proxmox/assets/151835671/c5474914-4655-4ba2-b6ae-6008ac942865)

  Proceed with manual build and install, as per documentation:
  <https://github.com/Telmate/terraform-provider-proxmox/blob/master/docs/guides/installation.md#manual-build--install>

  I used `make local-dev-install` and in `main.tf` source should be like this:

  ```hcl
  ...
  terraform {
    required_providers {
      proxmox = {
        source  = "localhost/telmate/proxmox"
        version = "3.0.2"
      }
    }
    required_version = ">= 1.8.4"
  }
  ...
  ```

  In `main.tf` define pool, which configured above:

  ```hcl
  ...
    pool         = "terraform-prov-pool"
  ...
  ```

- Run `terraform plan` and `terraform apply`. Now permission checking should ne passed and it is possible to create/destroy VMs in PVE, change VM parameters, add disks etc.

- So, for example, would be very nice and useful if will be such options in provider:
  
  - as wrote in previous post, make checking permissions functionality ( [Telmate/proxmox-api-go#308](https://github.com/Telmate/proxmox-api-go/issues/308)) optional? True/false or no permissions check for some user, as for example now is for root@pam.
  - possibility define permission checking for defined pool, not only for `/`
  - possibility define `minimumPermissions` list

**Comment by hestiahacker** [comment]
Error I'm getting from 3.0.1-rc3:

```
╷
│ Error: Invalid provider configuration
│ 
│ Provider "registry.terraform.io/telmate/proxmox" requires explicit configuration. Add a provider block to the root
│ module and configure the provider's required arguments as described in the provider documentation.
│ 
│
│ Error: user redacted@pam has valid credentials but cannot retrieve user list, check privilege separation of api token
│ 
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on <empty> line 0:
│   (source code not available)
│ 
╵
```

These were due to permission errors. Adding the `Administrators` role gets rid of these errors, but is way more permissions than is appropriate to give the account used for deployment.

Adding the `PVEUserAdmin` role gives a different error:

```
╷
│ Error: Invalid provider configuration
│ 
│ Provider "registry.terraform.io/telmate/proxmox" requires explicit configuration. Add a provider block to the root
│ module and configure the provider's required arguments as described in the provider documentation.
│ 
╵
╷
│ Error: Plugin did not respond
│ 
│   with provider["registry.terraform.io/telmate/proxmox"],
│   on <empty> line 0:
│   (source code not available)
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ConfigureProvider call. The
│ plugin logs may contain more details.
```

The key privilege here was `User.Modify`. That's what took us from "cannot retrieve user list" to "Plugin did not respond".

Long story short, here's a list of the additional permissions I needed from last time I reported in on this thread:

- User.Modify
- Sys.Audit
- Sys.Console (which is one I apparently identified earlier but my deployer user didn't have it)
- Sys.Modify (which is one I apparently identified earlier but my deployer user didn't have it)

Note: All of the roles were added to the root path (`/`) and had the Propagate checkbox checked.

**Comment by staticdev** [comment]
I am also using latest v3.0.1-rc6 and would be great to improve the `permissions_check` based on what needs to be done and minimum permissions so we avoid any security issues resulting for adding more than neeed.

**Comment by Avengerist** [comment]
> I am also using latest v3.0.1-rc6 and would be great to improve the `permissions_check` based on what needs to be done and minimum permissions so we avoid any security issues resulting from adding more than neeed. 

Unfortunately, the same problem, for two years I have not been able to provide the ability to use Proxmox to my users through terraform because of privileges. I hope the problem will be solved ASAP

**Comment by hestiahacker** [comment]
Just a heads up, `SDN.Use` should also be added to the documentation. This is needed even if you are not using SDN!

I'm not sure when it was added to the required permissions, but I can confirm that it is needed by 3.0.1-rc6.

There was a reasonably nice error message indicating what permission it was that was missing, so that's nice. However it's unfortunate that the software will not run without this permission when it should not be needed unless you are using SDN features.

**Comment by mattross-id-au** [comment]
I had the same error and solved it  by making sure I applied the correct permissions to the _Token_, not the User. 

```bash
pveum role add TerraformProv -privs "Datastore.AllocateSpace Datastore.AllocateTemplate Datastore.Audit Pool.Allocate Sys.Audit Sys.Console Sys.Modify VM.Allocate VM.Audit VM.Clone VM.Config.CDROM VM.Config.Cloudinit VM.Config.CPU VM.Config.Disk VM.Config.HWType VM.Config.Memory VM.Config.Network VM.Config.Options VM.Migrate VM.Monitor VM.PowerMgmt SDN.Use"
pveum user add terraform-prov@pve
pveum aclmod / -user terraform-prov@pve -role TerraformProv # Apply permission to the user
pveum user token add terraform-prov@pve api
pveum aclmod / --tokens 'terraform-prov@pve!api' -role TerraformProv # Apply permission to the token
```
